### DISCUSSION TITLE
BubbleUPNP player reports UPNP search not enabled on Gerbera Server

### DISCUSSION THREAD

**Discussion by BobManMasher** [gerbera/gerbera#2722] [Q&A]
I am currently running 1.11.0 on Debian Buster which I compilled when it was released earlier this year. I use Bubbleupnp as the client on my Android phones and I also use bubble server to give me remote access.  If I try to choose random albums or random tracks from my library, I always get a message pop up to say upnp search is not available on my server.  

Gerbera is compiled with the following options:
`cmake .. -DWITH_NPUPNP=YES -DWITH_JS=1 -DWITH_MYSQL=0 -DWITH_CURL=1 -DWITH_TAGLIB=1 -DWITH_MAGIC=1
-DWITH_MATROSKA=0 -DWITH_AVCODEC=1 -DWITH_EXIF=1 -DWITH_EXIV2=0 -DWITH_LASTFM=0
-DWITH_FFMPEGTHUMBNAILER=1 -DWITH_INOTIFY=1`

The server currently has 12,116 files available (circa 800 albums) and it connects beautifully to my phones, FireStick and other devices where I can select tracks or albums but with such a large number, a random selection stops me picking on the same things! I see there already is a similar query to this but they are using an old flavour of Gerbera.

I think I had this feature working many versions ago but now it's definitely not functioning.  Do I need to add anything to Config.xml or do I need a different compile time option?  Is there anyway I can test the UPNP Search function with using bubbleupnp in case the issue lies there?

Thanks - Bob

**Comment by KarlStraussberger** [comment]
I don't remember exactly, but there have been some flaws in 1.11 that may result in your setup. I just tried to search in my HEAD revision with bubble directly and got the result I expected.

Maybe you can run with `--debug` to get the search string which is not understood.

**Reply by BobManMasher** [reply]
Thanks Karl, I have attached the relevant section of the log file I produced.  
[upnpsearch.log](https://github.com/gerbera/gerbera/files/9847414/upnpsearch.log)
I will put my hands up - no idea whats going on and what it means but I believe the search request seems to be getting through and processed.  What should I be looking out for in the log? Could this be a problem with SQLite and would I be better off with something else like Mariadb?

**Reply by KarlStraussberger** [reply]
I have no clue, either. It runs a search for all albums with `RequestedCount 1`

```
2022-10-20 17:57:34.090   doSearch(): ContainerID [0] SearchCriteria [upnp:class = "object.container.album.musicAlbum"] StartingIndex [0] RequestedCount [1]
```
Get's what requested

```
<NumberReturned>1</NumberReturned>
<TotalMatches>9211</TotalMatches>
```

and then asks to continue at `StartingIndex 1300`.

```
2022-10-20 17:58:01.901  doSearch(): ContainerID [0] SearchCriteria [upnp:class = "object.container.album.musicAlbum"] StartingIndex [1300] RequestedCount [1]
```

Seems to be broken on the Bubble side of life.

**Reply by BobManMasher** [reply]
Thanks Karl, I raised the issue with the team at Bubble and supplied them logs from the client end. They responded as follows:

> The log file shows that UPnP search requests are emitted to Gerbera, but they fail on a timout of 30s, because Gerbera is not responding with data within 30s. Currently that timeout value is hard-coded so there is no way to test if it would return data with a higher value. Even if it did, it would make search very slow.

That is not to say that Gerbera is slow but more likely my implementation on low end plug servers is probably not great. On a more philosophical point, is there a faster database implementation than SQLite3 that I could use?

**Reply by KarlStraussberger** [reply]
I did have a closer look at your log and found the place also

```
2022-10-20 17:57:34.126  debug: [sql_database.cc:1017] search(): Search count resolves to SQL [SELECT COUNT(DISTINCT "c"."id") FROM "mt_cds_object" "c" INNER JOIN "mt_metadata" "m" ON "c"."id" = "m"."item_id" INNER JOIN "grb_cds_resource" "re" ON "c"."id" = "re"."item_id" WHERE (LOWER("c"."upnp_class")=LOWER('object.container.album.musicAlbum'))]
2022-10-20 17:57:34.127  debug: [sqlite_database.cc:229] select(): Adding select to Queue: SELECT COUNT(DISTINCT "c"."id") FROM "mt_cds_object" "c" INNER JOIN "mt_metadata" "m" ON "c"."id" = "m"."item_id" INNER JOIN "grb_cds_resource" "re" ON "c"."id" = "re"."item_id" WHERE (LOWER("c"."upnp_class")=LOWER('object.container.album.musicAlbum'))
2022-10-20 17:57:34.127  debug: [thread_runner.h:135] lockGuard(): ThreadRunner: Guard for SQLiteThread - addTask SelectTask
2022-10-20 17:57:34.127  debug: [thread_runner.h:125] notify(): ThreadRunner: Notifying SQLiteThread
2022-10-20 17:57:34.127  debug: [sqlite_database.cc:447] run(): Running: SELECT COUNT(DISTINCT "c"."id") FROM "mt_cds_object" "c" INNER JOIN "mt_metadata" "m" ON "c"."id" = "m"."item_id" INNER JOIN "grb_cds_resource" "re" ON "c"."id" = "re"."item_id" WHERE (LOWER("c"."upnp_class")=LOWER('object.container.album.musicAlbum'))
2022-10-20 17:57:58.241  debug: [thread_runner.h:95] wait(): ThreadRunner: Waiting for SQLiteThread
```

24 seconds to count the result size.

Trying to find the cause in the code I conclude that you are running 1.10 instead of 1.11 because in 1.11 this is already a "LEFT JOIN". So maybe you try to go to 1.11 or wait until tomorrow (I hope to be able to make a new release).

I tried with my HEAD and got a search for 1200+ albums in no time.

**Reply by BobManMasher** [reply]
Thanks Karl - that was a great spot.  I double checked my build files and I definitely built 1.11.0 on May 16th.  I also checked CMakeFiles.txt where I set the version numbers and both versions 1.10 and 1.11 report correctly.  Not quite sure whats going on there.  To avoid further confusion, it's probably safest for me to wait for the next version when you release it and try the new build.

**Comment by BobManMasher** [comment]
Hi Karl, Slightly off topic but related. Been trying to compile the new version (1.12.0) but had lots of problems. Now down to two warnings as follows:

`[ 57%] Building CXX object CMakeFiles/libgerbera.dir/src/metadata/taglib_handler.cc.o
/mnt/sata/build/gerbera-1.12.0/src/metadata/resolution.cc:36:25: warning: result of comparison of constant 18446744073709551615 with expression of type 'unsigned long' is always false [-Wtautological-constant-out-of-range-compare]
        if (x == 0 || x == std::numeric_limits<uint64_t>::max()) {
                      ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/sata/build/gerbera-1.12.0/src/metadata/resolution.cc:44:25: warning: result of comparison of constant 18446744073709551615 with expression of type 'unsigned long' is always false [-Wtautological-constant-out-of-range-compare]
        if (y == 0 || y == std::numeric_limits<uint64_t>::max()) {
                      ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.`

and the following error when linking which is fatal. Is the issue to do with the new version of FMT which I had to compile (latest version in the repo is 8 but 9 is the latest release version) or something in the build script? The Build logfile contains the following:

`# Run command: /usr/bin/cmake --build . --target "preinstall"'
'# Directory: /mnt/sata/build/gerbera-1.12.0/build'
'# Output:'
'[ 99%] Built target libgerbera
[100%] Linking CXX executable gerbera
/usr/bin/arm-linux-gnueabi-ld: CMakeFiles/gerbera.dir/src/main.cc.o: in function `fmt::v9::appender fmt::v9::vformat_to<fmt::v9::appender, 0>(fmt::v9::appender, fmt::v9::basic_string_view<char>, fmt::v9::basic_format_args<fmt::v9::basic_format_context<fmt::v9::appender, char> >)':
main.cc:(.text._ZN3fmt2v910vformat_toINS0_8appenderELi0EEET_S3_NS0_17basic_string_viewIcEENS0_17basic_format_argsINS0_20basic_format_contextIS2_cEEEE[_ZN3fmt2v910vformat_toINS0_8appenderELi0EEET_S3_NS0_17basic_string_viewIcEENS0_17basic_format_argsINS0_20basic_format_contextIS2_cEEEE]+0xf0): undefined reference to `_ZN3fmt2v96detail10vformat_toIcEEvRNS1_6bufferIT_EENS0_17basic_string_viewIS4_EENS0_17basic_format_argsINS0_20basic_format_contextINSt11conditionalIXsr3std7is_sameIS4_cEE5valueENS0_8appenderESt20back_insert_iteratorIS5_EE4typeES4_EEEENS1_10locale_refE'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [CMakeFiles/gerbera.dir/build.make:121: gerbera] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:97: CMakeFiles/gerbera.dir/all] Error 2
gmake: *** [Makefile:171: all] Error 2
`

I must crack cross compiling as it's taking around 12hrs at the moment which is still impressive given the hardware platform and resources it's running on. Ta - Bob

**Reply by KarlStraussberger** [reply]
The warnings are worth to look at.

The linker error might originate from multiple fmt versions on your system. Also, I remember some compatibility issues with fmt 9.0. Another reason can be too aggressive optimization of the compile which leads to dropped template symbols.

**Reply by KarlStraussberger** [reply]
Did you get more of integer size warnings. May be the root cause of some weird issues we have at the moment.

**Reply by BobManMasher** [reply]
Having a mini-mare at present with FMT. Thought I would wind everything, remove other occurrences etc. Now getting oddball errors compiling both ver 8.0.1 and 9.1.0. I realise as well that the repo only has libfmt and not FMT so it is way out of date and without that, I cant build spdlog which I need for gerbera. Will keep you posted and send any more Gerbera integer errors and try and get as much detail as I can once this is sorted - Bob
